File: matrix-traversals/bfs.java
int[][] dirs = { {0,1}, {0,-1}, {1,0}, {-1,0} };
void bfs(char[][] grid, int _i, int _j) {
    Deque<int[]> q = new ArrayDeque<>();
    boolean[][] visited = new boolean[grid.length][grid[0].length];
    q.addLast(new int[] {_i, _j});
    visited[_i][_j] = true;
    while (!q.isEmpty()) {
        int[] cur = q.removeFirst();
        for (int[] dir : dirs) {
            int i = cur[0] + dir[0];
            int j = cur[1] + dir[1];
            if (i < 0 || j < 0 || i >= grid.length || 
                j >= grid[0].length || visited[i][j]) continue;
            visited[i][j] = true;
            q.addLast(new int[] {i, j});
        }
    }
}

File: matrix-traversals/bfs.py
dirs = [ [0,1], [0,-1], [1,0], [-1,0] ]

def bfs(grid: List[List[int]], _i: int, _j: int):
    q = deque([[_i, _j]])
    visited = set([(_i, _j)])
    
    while q:
        cur = q.popleft()

        for dir in dirs:
            i = cur[0] + dir[0]
            j = cur[1] + dir[1]
            
            if i < 0 or j < 0 or i >= len(grid) or \
            j >= len(grid[0]) or (i, j) in visited: continue
            
            visited.add((i, j))
            q.append([i, j])